src/lib/update-lock.ts

```diff
@@ -82,22 +90,188 @@ async function dispose(
+		await dispose(appId, release);
+		// Re-throw error to be handled in caller
+		throw err;
+	} finally {
+		// If not already released from catch, released the RW process lock.
+		// If already released, this will not error.
+		release();
+	}
+}
+
+/**
+ * Composition step used by Supervisor compose module.
+ * Release all locks for an appId | appUuid.
+ */
+export async function releaseLock(appId: number) {
+	logger.logSystemEvent(logTypes.releaseLock, { appId });
+
+	const release = await takeGlobalLockRW(appId);
+	await dispose(appId, release);
+}
+
+/**
+ * Given a lockfile path `p`, return an array [appId, serviceName, filename] of that path.
+ * Paths are assumed to end in the format /:appId/:serviceName/(resin-)updates.lock.
+ */
+function getIdentifiersFromPath(p: string) {
+	const parts = p.split('/');
+	const filename = parts.pop();
+	if (filename?.match(/updates\.lock/) === null) {
+		return [];
+	}
+	const serviceName = parts.pop();
+	const appId = parts.pop();
+	return [appId, serviceName, filename];
+}
+
+type LockedEntity = { appId: number; services: string[] };
+
+/**
+ * A map of locked services by appId.
+ * Exported for tests only; getServicesLockedByAppId is the public generator interface.
+ */
+export class LocksTakenMap extends Map<number, Set<string>> {
+	constructor(lockedEntities: LockedEntity[] = []) {
+		// Construct a Map<number, Set<string>> from user-friendly input args
+		super(
+			lockedEntities.map(({ appId, services }) => [appId, new Set(services)]),
+		);
+	}
+
+	// Add one or more locked services to an appId
+	public add(appId: number, services: string | string[]): void {
+		if (typeof services === 'string') {
+			services = [services];
+		}
+		if (this.has(appId)) {
+			const lockedSvcs = this.get(appId)!;
+			services.forEach((s) => lockedSvcs.add(s));
+		} else {
+			this.set(appId, new Set(services));
+		}
+	}
+
+	/**
+	 * @private Use this.getServices instead as there is no need to return
+	 * a mutable reference to the internal Set data structure.
+	 */
+	public get(appId: number): Set<string> | undefined {
+		return super.get(appId);
+	}
+
+	// Return an array copy of locked services under an appId
+	public getServices(appId: number): string[] {
+		return this.has(appId) ? Array.from(this.get(appId)!) : [];
+	}
+
+	// Return whether a service is locked under an appId
+	public isLocked(appId: number, service: string): boolean {
+		return this.has(appId) && this.get(appId)!.has(service);
+	}
+}
+
+// A wrapper function for lockfile.getLocksTaken.
+// Exported for tests only; getServicesLockedByAppId is the intended public interface.
+export const getLocksTaken = lockfile.getLocksTaken;
+
+/**
+ * Return a list of services that are locked by the Supervisor under each appId.
+ * Both `resin-updates.lock` and `updates.lock` should be present per
+ * [appId, serviceName] pair for a service to be considered locked.
+ */
+export async function getServicesLockedByAppId(): Promise<LocksTakenMap> {
```

Comment:
This doesn't need to be async now

Reply:
Reverted the in-memory store, so the function is async again